File: packages/gavio-js/src/interceptors/pii/scanners/credit-card.ts
/** Credit card scanner — regex candidate + Luhn checksum validation. */

import type { ScanContext } from '../context.js'
import { makeMatch } from '../match.js'
import type { PiiMatch } from '../match.js'
import type { PiiScanner } from '../scanner.js'

// 13–19 digits, optionally separated by single spaces or hyphens.
const CARD = /\b(?:\d[ -]?){12,18}\d\b/g

export function luhnValid(number: string): boolean {
  const digits: number[] = []
  for (const c of number) {
    if (c >= '0' && c <= '9') digits.push(c.charCodeAt(0) - 48)
  }
  if (digits.length < 13 || digits.length > 19) return false
  let checksum = 0
  const parity = digits.length % 2
  for (let i = 0; i < digits.length; i++) {
    let d = digits[i]!
    if (i % 2 === parity) {
      d *= 2
      if (d > 9) d -= 9
    }
    checksum += d
  }
  return checksum % 10 === 0
}

export function creditCardScanner(): PiiScanner {
  return {
    entityType: 'CREDIT_CARD',
    tier: 1,
    scan(text: string, ctx: ScanContext): PiiMatch[] {
      const out: PiiMatch[] = []
      for (const m of text.matchAll(CARD)) {
        if (!luhnValid(m[0])) continue
        const idx = ctx.nextIndex('CREDIT_CARD')
        out.push(
          makeMatch({
            entityType: 'CREDIT_CARD',
            start: m.index,
            end: m.index + m[0].length,
            value: m[0],
            replacement: `[CREDIT_CARD_${idx}]`,
          }),
        )
      }
      return out
    },
  }
}

File: packages/gavio-py/gavio/interceptors/pii/scanners/credit_card.py
"""Credit card scanner — regex candidate + Luhn checksum validation."""

from __future__ import annotations

import re

from ..context import ScanContext
from ..match import PiiMatch
from ..scanner import PiiScanner

# 13–19 digits, optionally separated by single spaces or hyphens.
_CARD = re.compile(r"\b(?:\d[ -]?){12,18}\d\b")


def _luhn_valid(number: str) -> bool:
    digits = [int(c) for c in number if c.isdigit()]
    if not 13 <= len(digits) <= 19:
        return False
    checksum = 0
    parity = len(digits) % 2
    for i, d in enumerate(digits):
        if i % 2 == parity:
            d *= 2
            if d > 9:
                d -= 9
        checksum += d
    return checksum % 10 == 0


class CreditCardScanner(PiiScanner):
    entity_type = "CREDIT_CARD"

    def scan(self, text: str, ctx: ScanContext) -> list[PiiMatch]:
        out: list[PiiMatch] = []
        for m in _CARD.finditer(text):
            if not _luhn_valid(m.group()):
                continue
            idx = ctx.next_index(self.entity_type)
            out.append(
                PiiMatch(
                    entity_type=self.entity_type,
                    start=m.start(),
                    end=m.end(),
                    value=m.group(),
                    replacement=f"[CREDIT_CARD_{idx}]",
                )
            )
        return out
